ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -0,0 +1,545 @@
+  | LIInitialiseFromLMDB FS.FsPath -- ^ Initialise by copying from an LMDB db at this path
+  | LINoInitialise -- ^ The database is already initialised
+
+-- | Initialize the LMDB from these provided values
+initFromVals ::
+     TableStuff l
+  => WithOrigin SlotNo
+  -> LedgerTables l ValuesMK
+  -> Db l
+  -> IO ()
+initFromVals dbsSeq vals Db{..} = LMDB.readWriteTransaction dbEnv $
+  withDbSettingsRWMaybeNull dbSettings $ \case
+    Nothing -> zipLedgerTablesA initLMDBTable dbBackingTables vals $> ((), DbState{dbsSeq})
+    Just _ -> Exn.throw $ DbErrStr "initFromVals: db already had state"
+
+initFromLMDBs :: MonadIO m
+  => Trace.Tracer IO TraceDb
+  -> FS.SomeHasFS m
+  -> FS.FsPath
+  -> FS.FsPath
+  -> m ()
+initFromLMDBs tracer shfs from0 to0 = do
+  liftIO $ Trace.traceWith tracer $ TDBInitialisingFromLMDB from0
+  from <- guardDbDir GDDMustExist shfs from0
+  to <- guardDbDir GDDMustNotExist shfs to0
+  liftIO $ Exn.bracket
+    (LMDB.openEnvironment from defaultLMDBLimits) -- TODO assess limits, in particular this could fail if the db is too big
+    LMDB.closeEnvironment $ \e -> lmdbCopy tracer e to
+
+
+lmdbCopy :: MonadIO m
+  => Trace.Tracer m TraceDb
+  -> LMDB.Environment LMDB.ReadWrite
+  -> FilePath
+  -> m ()
+lmdbCopy tracer dbEnv to = do
+  -- TODO This copying is gross. Tests depend on it, but I wish I was only responsible for copying the database here
+    let LMDB.Env e = dbEnv
+    from <- liftIO $ LMDB.mdb_env_get_path e
+    Trace.traceWith tracer $ TDBCopying from to
+    liftIO $ LMDB.mdb_env_copy e to
+    -- others <- liftIO $ Dir.listDirectory from <&> filter ((/= ".mdb") . Dir.takeExtension)
+    -- for_ others $ \f0 -> do
+    --   let f = from Dir.</> f0
+    --       t = to Dir.</> f0
+    --   Trace.traceWith tracer $ TDBCopyingFile f t
+    --   liftIO $ Dir.copyFile f t
+    Trace.traceWith tracer $ TDBCopied from to
+
+-- TODO We don't know how much time is spent in I/O
+
+-- TODO 50% of total time is spent somewhere else. Where?
+
+-- | Initialise a backing store
+newLMDBBackingStore :: forall m l. (TableStuff l, MonadIO m, IOLike m)
+  => Trace.Tracer m TraceDb
+  -> LMDBLimits
+  -> FS.SomeHasFS m
+  -> LMDBInit l
+  -> m (LMDBBackingStore l m)
+newLMDBBackingStore tracer limits sfs init_db = do
+  let
+    path = FS.mkFsPath ["tables"]
+  Trace.traceWith tracer TDBOpening
+
+  -- Operations invoked on the returned LMDBBackingStore end up running in IO
+  -- Since the tracer we were passed runs in m, we use this async to match the
+  -- Tracer IO we want to use with the Tracer m we have
+  -- TODO exception safety, we should trace_async 'onException'
+  --
+
+  -- TODO why are we using TBQueues?? is it really needed?
+
+  -- TODO This tracing logic also could use a review (perhaps use `natTracer` to transform the tracer?)
+  trace_q <- liftIO $ IOLike.newTBQueueIO 1
+  trace_async <- IOLike.async . void . runMaybeT . forever $ do
```

Comment:
In addition to the comment above I wonder why do we need to trace asynchronously.

Reply:
After reviewing the uses of `dbTracer` in this module I think we should be  able to get rid of this additional complexity.